odbc/src/api/diag.rs

```diff
@@ -0,0 +1,159 @@
+use crate::{
+    api::{data::i16_len, definitions::DiagType},
+    errors::ODBCError,
+};
+use odbc_sys::Pointer;
+use odbc_sys::{Char, Integer, SmallInt, SqlReturn, WChar};
+use std::ptr::copy_nonoverlapping;
+
+///
+/// set_sql_state writes the given sql state to the [`output_ptr`].
+///
+/// # Safety
+/// This writes to a raw C-pointer
+///
+pub unsafe fn set_sql_state(sql_state: &str, output_ptr: *mut WChar) {
+    if output_ptr.is_null() {
+        return;
+    }
+    let sql_state = &format!("{}\0", sql_state);
+    let state_u16 = sql_state.encode_utf16().collect::<Vec<u16>>();
+    copy_nonoverlapping(state_u16.as_ptr(), output_ptr, 6);
+}
+
+///
+/// get_diag_rec copies the given ODBC error's diagnostic information
+/// into the provided pointers.
+///
+/// # Safety
+/// This writes to multiple raw C-pointers
+///
+pub unsafe fn get_diag_rec(
+    error: &ODBCError,
+    state: *mut WChar,
+    message_text: Pointer,
+    buffer_length: SmallInt,
+    text_length_ptr: *mut SmallInt,
+    native_error_ptr: *mut Integer,
+    wstring: bool,
+) -> SqlReturn {
+    if !native_error_ptr.is_null() {
+        *native_error_ptr = error.get_native_err_code();
+    }
+    set_sql_state(error.get_sql_state(), state);
+    let message = format!("{}", error);
+    match wstring {
+        true => i16_len::set_output_wstring(
+            &message,
+            message_text as *mut WChar,
+            buffer_length as usize,
+            text_length_ptr,
+        ),
+        false => i16_len::set_output_string(
+            &message,
+            message_text as *mut Char,
+            buffer_length as usize,
+            text_length_ptr,
+        ),
+    }
+}
+
+///
+/// get_stmt_diag_field copies a part of the given ODBC error's diagnostic information
+/// into the provided pointer.
+///
+/// # Safety
+/// This writes to a raw C-pointer
+///
+pub unsafe fn get_stmt_diag_field(diag_identifier: DiagType, diag_info_ptr: Pointer) -> SqlReturn {
+    // NOTE: at the moment, this could be merged with get_diag_field. However, as part of SQL-1152,
+    // some functionality will be specific to the statement handle, and thus warrants a separate function
+    match diag_identifier {
+        // default to 0, mirroring the behavior in SQLRowCount
+        DiagType::SQL_DIAG_ROW_COUNT => {
+            i16_len::set_output_fixed_data(&0isize, diag_info_ptr, &mut 0)
+        }
+        DiagType::SQL_DIAG_ROW_NUMBER => {
+            // default to unknown, as at the moment statement handels don't update their row number attribute
+            const SQL_ROW_NUMBER_UNKNOWN: isize = -2;
```

Comment:
I would consider moving it to `odbc/src/api/definitions.rs`. This way you can also use it in the test when validating the result.